pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessagesImpl.java

```diff
@@ -0,0 +1,78 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.client.impl;
+
+import com.google.common.base.Preconditions;
+import org.apache.pulsar.client.api.Message;
+import org.apache.pulsar.client.api.Messages;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+
+public class MessagesImpl<T> implements Messages<T> {
+
+    private List<Message<T>> messageList;
+
+    private final int maxNumberOfMessages;
+    private final long maxSizeOfMessages;
+
+    private int currentNumberOfMessages;
+    private long currentSizeOfMessages;
+
+    protected MessagesImpl(int maxNumberOfMessages, long maxSizeOfMessages) {
+        this.maxNumberOfMessages = maxNumberOfMessages;
+        this.maxSizeOfMessages = maxSizeOfMessages;
+        messageList = maxNumberOfMessages > 0 ? new ArrayList<>(maxNumberOfMessages) : new ArrayList<>();
+    }
+
+    protected boolean canAdd(Message<T> message) {
+        if (maxNumberOfMessages <= 0 && maxSizeOfMessages <= 0) {
+            return true;
+        }
+        return (maxNumberOfMessages > 0 && currentNumberOfMessages + 1 <= maxNumberOfMessages)
+                || (maxSizeOfMessages > 0 && currentSizeOfMessages + message.getData().length <= maxSizeOfMessages);
+    }
+
+    protected void add(Message<T> message) {
+        if (message == null) {
+            return;
+        }
+        Preconditions.checkArgument(canAdd(message), "No more space to add messages.");
```

Comment:
maybe just change this method to return `boolean` and make `canAdd(message)` to a private method.

```
protected boolean add(Message<T> message);
```

Reply:
We need canAdd to do a pre-check, because we need add intercepted messages to Messages.